### PR TITLE
WIP: Register node-inspector as a RemoteDebug-endpoint over mDNS/Bonjour

### PR DESCRIPTION
**Pull request by auchenberg**
This PR makes it possible for external tools like [chrome-devtools-app](https://github.com/auchenberg/chrome-devtools-app/), to discover an instance of node-inspector as a "RemoteDebug-endpoint" over mDNS/Bonjour, as a follow up on https://github.com/node-inspector/node-inspector/issues/506. 

The idea here is to enable discovery of [RemoteDebug](http://remotedebug.org)-compliant endpoints over MDNS. The metadata is based upon the existing Chrome Remote Debug HTTP API. I'm planning to write a little spec for RemoteDebug Endpoint MDNS functionality too. 

bonjour browser
### Changes:
- [x] New utility methods has been added to generate URL's.
- [x] Added remoteDebugService to register the service over MDNS.
- [x] Change default web-host to local detected IP instead of 127.0.0.1
### Questions:

1) I had to change the default web-host to the local IP instead of 127.0.0.1, as the IP is needed in order to enable other clients to connect. Does this open up a security risk? Is there any reason why 127.0.0.1/0.0.0.0 currently is used?

### CODE REVIEW

**Comment by 3y3** [comment]
@bajtos , please join to discussion.

@auchenberg , thanks for contribution. This is interesting feature.

>  Is there any reason why 127.0.0.1/0.0.0.0 currently is used?

From time to time I know why we use `0.0.0.0` in `NODE_INSPECTOR_MODE` and `127.0.0.1` in `NODE_DEBUG_MODE`, but unfortunately now I can't remember this.

> I had to change the default web-host to the local IP instead of 127.0.0.1, as the IP is needed in order to enable other clients to connect.

Can you explain me more verbose this part. At first glance new IP detecting opens weak security risk. I prefer to have minimal network area in default configuration, but it's discussable.

File: lib/config.js [hunk 1]
```diff
@@ -435,7 +436,6 @@ function collectDefaultsFromDefinitions(NODE_DEBUG_MODE) {
   });
 
   if (NODE_DEBUG_MODE) {
-    options['web-host'] = '127.0.0.1';
```
Comment (bajtos) [inline lib/config.js]: Please keep `127.0.0.1` for `NODE_DEBUG_MODE`. This mode is triggered when debugging via `node-debug app.js`, in which case the front-end is immediately opened in the browser.

File: package.json [hunk 2]
```diff
@@ -35,7 +35,9 @@
     "which": "^1.0.5",
     "v8-debug": "~0.3.0",
     "v8-profiler": "~5.2.0",
-    "semver": "^3.0.1"
+    "semver": "^3.0.1",
+    "mdns": "~2.2.2",
```
Comment (bajtos) [inline package.json]: Please use `^2.2.2`.

**Comment by bajtos** [comment]
> I had to change the default web-host to the local IP instead of 127.0.0.1, as the IP is needed in order to enable other clients to connect. Does this open up a security risk? Is there any reason why 127.0.0.1/0.0.0.0 currently is used?

127.0.0.1 is used in node-debug mode, as it is the most sensible value. The user is loading the frontend in a local browser, thus a local loopback address is good enough for making the connection, and there is no need to open the debugger server to the whole network.

When running `node-inspector` in a standalone mode, 0.0.0.0 is used mostly for historical reasons. I don't mind changing it to 127.0.0.1, as that is more secure. However, I am opposed to changing it to network IP address as proposed here.

IIUC mDNS/Bonjour correctly, "it provides a general method to discover services on a local area network" ([wikipedia](http://en.wikipedia.org/wiki/Bonjour_%28software%29)), which means that we need to register with an IP address visible from the local network.

I am proposing the following changes to your patch:
 - Keep the configuration of listening address intact, i.e. 127.0.0.1/0.0.0.0.
 - Modify the code registering with mDNS and make it smarter. Detect the situation when the server is listening on 0.0.0.0 and replace 0.0.0.0 with the local network address.

**Comment by auchenberg** [comment]
Good points @bajtos. 

My intention is to enable the MDNS broadcasting out of the box, in order to minimize the workflow, but I understand the added security risk of binding to a addressable IP. 

Alternatively we could add a command-line flag that would enable MDNS broadcasting, the user could flip when running ``node-debug`` and ``node-inspector``? That way it's explicit. 

Flag enabled: ``web-host`` is set to the local addressable IP, and MDNS broadcasting is enabled

**Comment by bajtos** [comment]
I am ok with enabling MDNS broadcasting always, or alternatively when Node Inspector is listening on a public address.

Here is what I am proposing:
 - `node-debug app.js` keeps listening on 127.0.0.1, this command is not intended for debugging via chrome-devtools-app.
 - `node-inspector app.js` keeps listening on 0.0.0.0, a public address is registered with MDNS.

**Comment by auchenberg** [comment]
Changes implemented:

- [x] node-debug app.js keeps listening on ``127.0.0.1``, this command is not intended for debugging via chrome-devtools-app.

- [x] When ``web-host`` is set to ``0.0.0.0``, the local addressable IP is detected, and used for websocket, inspector and frontend URL generation, plus MDNS broadcast is enabled. 

I still have 3 tests failing, but these are also failing for me in mester. Any cure?

**Comment by 3y3** [comment]
Are you using node v0.10.35?

**Comment by auchenberg** [comment]
I had 3 tests failing on v0.10.33, and now with 0.10.35, 29 tests are failing.

File: lib/debug-server.js [hunk 3]
```diff
@@ -78,6 +94,11 @@ DebugServer.prototype._createSession = function(debugPort) {
   return Session.create(debugPort, this._config);
 };
 
+DebugServer.prototype._shouldAnnounceEndpoint = function() {
+  var addressInfo = this._httpServer.address();
+  return addressInfo.address === '0.0.0.0';
```
Comment (bajtos) [inline lib/debug-server.js]: I think we should support any non-localhost IPs too.

http://tools.ietf.org/html/rfc5735
```
127.0.0.0/8 - This block is assigned for use as the Internet host
loopback address.  A datagram sent by a higher-level protocol to an
address anywhere within this block loops back inside the host.  This
is ordinarily implemented using only 127.0.0.1/32 for loopback.  As
described in [RFC1122], Section 3.2.1.3, addresses within the entire
127.0.0.0/8 block do not legitimately appear on any network anywhere.
```

A proposed condition:

```js
return !/^127\.0\.0\./.test(addressInfo.address)
```

Please include a descriptive comment linking to the RFC.

File: lib/debug-server.js [hunk 4]
```diff
@@ -85,11 +106,40 @@ DebugServer.prototype.close = function() {
...
 };
 
-DebugServer.prototype.address = function() {
-  var address = this._httpServer.address();
+DebugServer.prototype.getHttpServerAddressInfo = function() {
+  var addressInfo = this._httpServer.address();
+
+  if(this._shouldAnnounceEndpoint()) {
+    // Resolve local IP, in order for computers in the network to resolve
+    addressInfo.address = localIp();
```
Comment (bajtos) [inline lib/debug-server.js]: Local IP should be used only when the address is `0.0.0.0`. When the user explicitly specifies a public IP, then this IP should be honoured in the broadcasted event.

**Comment by auchenberg** [comment]
I'm holding on this PR, until I find a suitable replacement for ``mdns``, as we can't add such a dependency to the project.

**Comment by bajtos** [comment]
Closing as abandoned.

**Pull request closed without merging**